File: code_snippets/java/src/main/java/guides/RetryRunService.java
package guides;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Accept;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Raw;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.common.RetryPolicy;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.common.TimeoutException;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import develop.MyServiceClient;
import java.time.Duration;

@Service
public class RetryRunService {

  @Handler
  public String myRunHandler(Context ctx, String greeting) {

    // <start_here>
    // <mark_1>
    RetryPolicy myRunRetryPolicy =
        RetryPolicy.exponential(Duration.ofMillis(500), 2)
            .setMaxDelay(Duration.ofSeconds(10))
            .setMaxAttempts(10)
            .setMaxDuration(Duration.ofMinutes(5));
    // </mark_1>
    ctx.run("my-run", myRunRetryPolicy, () -> writeToOtherSystem());
    // <end_here>

    // <start_catch>
    try {
      // Fails with a terminal error after 3 attempts or if the function throws one
      ctx.run("my-run", RetryPolicy.defaultPolicy().setMaxAttempts(3), () -> writeToOtherSystem());
    } catch (TerminalException e) {
      // Handle the terminal error: undo previous actions and
      // propagate the error back to the caller
    }
    // <end_catch>

    return greeting + "!";
  }

  // <start_raw>
  @Handler
  public void myHandler(
      Context ctx,
      // !mark
      @Accept("*/*") @Raw byte[] request) {
    try {
      var decodedRequest = decodeRequest(request);

      // ... rest of your business logic ...

    } catch (TerminalException e) {
      // Propagate to DLQ/catch-all handler
    }
  }

  // <end_raw>

  @Handler
  public void timeoutHandler(Context ctx) {
    // <start_timeout>
    try {
      // If the timeout hits first, it throws a `TimeoutError`.
      // If you do not catch it, it will lead to a retry.
      MyServiceClient.fromContext(ctx)
          .myHandler("Hello")
          // !mark
          .await(Duration.ofSeconds(5));

      var awakeable = ctx.awakeable(Boolean.class);
      // ...Do something that will trigger the awakeable
      // !mark
      awakeable.await(Duration.ofSeconds(5));

    } catch (TimeoutException e) {
      // Handle the timeout error
    }
    // <end_timeout>
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new RetryRunService()));
  }

  private Object decodeRequest(byte[] request) {
    return new Object();
  }

  private void writeToOtherSystem() {}
}

File: code_snippets/kotlin/src/main/kotlin/guides/RetryRunService.kt
package guides

import dev.restate.sdk.annotation.Accept
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Raw
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import dev.restate.sdk.common.TerminalException
import develop.MyServiceClient
import java.util.concurrent.TimeoutException
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

@Service
class RetryRunService {

  @Handler
  suspend fun retryRun(ctx: Context, greeting: String): String {

    // <start_here>
    // <mark_1>
    val myRunRetryPolicy = retryPolicy {
      initialDelay = 5.seconds
      exponentiationFactor = 2.0f
      maxDelay = 60.seconds
      maxAttempts = 10
      maxDuration = 5.minutes
    }
    // </mark_1>
    ctx.runBlock("write", myRunRetryPolicy) { writeToOtherSystem() }
    // <end_here>

    // <start_catch>
    try {
      // Fails with a terminal error after 3 attempts or if the function throws one
      ctx.runBlock(
          "write",
          RetryPolicy(
              initialDelay = 500.milliseconds, maxAttempts = 3, exponentiationFactor = 2.0f)) {
            writeToOtherSystem()
          }
    } catch (e: TerminalException) {
      // Handle the terminal error: undo previous actions and
      // propagate the error back to the caller
      throw e
    }
    // <end_catch>

    return "$greeting!"
  }

  // <start_raw>
  @Handler
  suspend fun myHandler(
      ctx: Context,
      // !mark
      @Accept("*/*") @Raw request: ByteArray
  ) {
    try {
      val decodedRequest = decodeRequest(request)

      // ... rest of your business logic ...

    } catch (e: TerminalException) {
      // Propagate to DLQ/catch-all handler
      throw e
    }
  }
  // <end_raw>

  @Handler
  suspend fun myTimeoutHandler(ctx: Context) {
    // <start_timeout>
    val awakeable = ctx.awakeable<String>()
    // do something that will trigger the awakeable
    // !mark(1:5
    val timeout = ctx.timer(5.seconds)
    try {
      val result = select {
        awakeable.onAwait { it }
        // !mark
        timeout.onAwait { throw TimeoutException() }
      }
    } catch (e: TimeoutException) {
      // Handle the timeout
    }

    val callFuture = MyServiceClient.fromContext(ctx).myHandler("Hello")
    // !mark
    val callTimeout = ctx.timer(5.seconds)
    try {
      val result = select {
        callFuture.onAwait { it }
        // !mark
        callTimeout.onAwait { throw TimeoutException() }
      }
    } catch (e: TimeoutException) {
      // Handle the timeout
    }
    // ... rest of your business logic ...
    // <end_timeout>
  }

  private fun decodeRequest(value: Any): String {
    TODO("Not yet implemented")
    return ""
  }

  private fun writeToOtherSystem(): String {
    return "writeToOtherSystem"
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(RetryRunService()) })
}
